cpp/tests/utilities/compare_column_wrappers.cuh

```diff
@@ -0,0 +1,146 @@
+}
+
+// TODO: Make sure the above function does something reasonable for all wrapped types
+
+} // namespace detail
+
+enum : bool {
+    distinct_nulls_are_equal = true,
+    distinct_nulls_are_unequal = false,
+};
+
+enum : bool {
+    do_print_unequal_pairs = true,
+    dont_print_unequal_pairs = false,
+};
+
+/**
+ * @note Currently ignoring the extra type info, i.e. assuming it's the same for both columns, or can be ignored
+ */
+template<typename E>
+void expect_column_values_are_equal(
+    gdf_size_type          common_size,
+    const E*               lhs_data_on_host,
+    const gdf_valid_type*  lhs_validity_on_host,
+    const std::string&     lhs_name,
+    const E*               rhs_data_on_host,
+    const gdf_valid_type*  rhs_validity_on_host,
+    const std::string&     rhs_name,
+    bool                   treat_distinct_nulls_as_equal = distinct_nulls_are_equal,
+    bool                   print_all_unequal_pairs = dont_print_unequal_pairs)
+{
+    auto lhs_non_nullable = (lhs_validity_on_host == nullptr);
+    auto rhs_non_nullable = (rhs_validity_on_host == nullptr);
+    auto max_name_length = std::max(lhs_name.length(), rhs_name.length());
+
+    for(gdf_size_type i = 0; i < common_size; i++) {
+        auto lhs_element_is_valid = lhs_non_nullable or gdf::util::bit_is_set<gdf_valid_type, gdf_size_type>(lhs_validity_on_host, i);
+        auto rhs_element_is_valid = rhs_non_nullable or gdf::util::bit_is_set<gdf_valid_type, gdf_size_type>(rhs_validity_on_host, i);
+        auto elements_are_equal =
+            (treat_distinct_nulls_as_equal and not lhs_element_is_valid and not rhs_element_is_valid) or
+            (lhs_element_is_valid == rhs_element_is_valid and lhs_data_on_host[i] == rhs_data_on_host[i]);
+        EXPECT_TRUE(elements_are_equal)
+            << std::left << std::setw(max_name_length) << lhs_name << std::right << '[' << i << "] = " << (lhs_element_is_valid ? detail::to_string(lhs_data_on_host[i]) : "@") << '\n'
+            << std::left << std::setw(max_name_length) << rhs_name << std::right << '[' << i << "] = " << (rhs_element_is_valid ? detail::to_string(rhs_data_on_host[i]) : "@") ;
+        if (not print_all_unequal_pairs and not elements_are_equal) { break; }
+    }
+}
+
+
+template<typename E>
+void expect_columns_are_equal(
+    cudf::test::column_wrapper<E> const&  lhs,
+    const std::string&                    lhs_name,
+    cudf::test::column_wrapper<E> const&  rhs,
+    const std::string&                    rhs_name,
+    bool                                  treat_distinct_nulls_as_equal = distinct_nulls_are_equal,
+    bool                                  print_all_unequal_pairs = dont_print_unequal_pairs)
+{
+    const gdf_column& lhs_gdf_column = *(lhs.get());
+    const gdf_column& rhs_gdf_column = *(rhs.get());
+    cudf::validate(lhs_gdf_column);
+    cudf::validate(rhs_gdf_column);
+    EXPECT_TRUE(cudf::have_same_type(lhs_gdf_column, rhs_gdf_column));
```

Comment:
I think it'd be nice to know _why_ two columns don't have the same type, and so I'd move the body of `have_same_type` to here with proper `EXPECT_EQ`, e.g.:
```
    EXPECT_EQ(lhs_gdf_column.dtype, rhs_gdf_column.dtype);
    EXPECT_EQ(is_nullable(lhs_gdf_column), is_nullable(rhs_gdf_column));
    if (lhs_gdf_column.dtype == GDF_TIMESTAMP) {
      EXPECT_EQ(lhs_gdf_column.dtype_info.time_unit,
                lhs_gdf_column.dtype_info.time_unit);
    }
```

Reply:
Maybe I'll just write an `expect_same_type()` function then.

Reply:
How about:
```
// Note: This has quite a bit of code repretition, plus it repeats all of `cudf::util::have_same_type` essentially.
bool expect_columns_have_same_type(const gdf_column& validated_column_1, const gdf_column& validated_column_2, bool ignore_extra_type_info = false) noexcept
{
    EXPECT_EQ(validated_column_1.dtype, validated_column_2.dtype);
    if(validated_column_1.dtype != validated_column_2.dtype) { return false; }
    EXPECT_EQ(cudf::is_nullable(validated_column_1), cudf::is_nullable(validated_column_2));
    if (cudf::is_nullable(validated_column_1) != cudf::is_nullable(validated_column_2)) { return false; }
    if (ignore_extra_type_info) { return true; }
    auto common_dtype = validated_column_1.dtype;
    EXPECT_TRUE(cudf::detail::extra_type_info_is_compatible(common_dtype, validated_column_1.dtype_info, validated_column_2.dtype_info));
    return cudf::detail::extra_type_info_is_compatible(common_dtype, validated_column_1.dtype_info, validated_column_2.dtype_info);
}
```

Reply:
That looks great.